broker/cloud_functions/ps_to_gcs/main.py

```diff
@@ -0,0 +1,169 @@
+
+    # Get the survey name and version
+    survey = guess_schema_survey(data)
+    version = guess_schema_version(data)
+
+    # By default, spool data in memory to avoid IO unless data is too big
+    # LSST alerts are anticipated at 80 kB, so 150 kB should be plenty
+    max_alert_packet_size = 150000
+    with TempAlertFile(max_size=max_alert_packet_size, mode='w+b') as temp_file:
+        temp_file.write(data)
+        temp_file.seek(0)
+        fix_schema(temp_file, survey, version, filename)
+        blob.upload_from_file(temp_file)
+        logger.log_text(f'Uploaded {filename} to {bucket.name}')
+
+def fix_schema(temp_file, survey, version, filename):
+    """ Rewrites the temp_file with a corrected schema header
+        so that it is valid for upload to BigQuery.
+
+    Args:
+        temp_file: Temporary file containing the alert.
+        survey: Name of the survey generating the alert.
+        version: Schema version.
+    """
+
+    # get the corrected schema if it exists, else return
+    try:
+        fpkl = f'valid_schemas/{survey}_v{version}.pkl'
+        inpath = Path(__file__).resolve().parent / fpkl
+        with inpath.open('rb') as infile:
+            valid_schema = pickle.load(infile)
+
+    except FileNotFoundError:
+        msg = f'Original schema header retained for {survey} v{version}; file {filename}'
+        logger.log_text(msg)
+        return
+
+    # load the file and get the data with fastavro
+    temp_file.seek(0)
+    data = [r for r in fastavro.reader(temp_file)]
+
+    # write the corrected file
+    temp_file.seek(0)
+    fastavro.writer(temp_file, valid_schema, data)
+    temp_file.truncate()  # removes leftover data
+    temp_file.seek(0)
+
+    logger.log_text(f'Schema header reformatted for {survey} v{version}; file {filename}')
+
+def guess_schema_version(alert_bytes: bytes) -> str:
+    """Retrieve the ZTF schema version
+
+    Args:
+        alert_bytes: An alert from ZTF or LSST
+
+    Returns:
+        The schema version
+    """
+
+    version_regex_pattern = b'("version":\s")([0-9]*\.[0-9]*)(")'
+    version_match = re.search(version_regex_pattern, alert_bytes)
+    if version_match is None:
+        err_msg = f'Could not guess schema version for alert {alert_bytes}'
+        logger.log_text(err_msg, severity='ERROR')
+        raise SchemaParsingError(err_msg)
+
+    return version_match.group(2).decode()
+
+def guess_schema_survey(alert_bytes: bytes) -> str:
+    """Retrieve the ZTF schema version
+
+    Args:
+        alert_bytes: An alert from ZTF or LSST
+
+    Returns:
+        The survey name
+    """
+
+    survey_regex_pattern = b'("namespace":\s")(\S*)(")'
+    survey_match = re.search(survey_regex_pattern, alert_bytes)
+    if survey_match is None:
+        err_msg = f'Could not guess survey name for alert {alert_bytes}'
+        logger.log_text(err_msg, severity='ERROR')
+        raise SchemaParsingError(err_msg)
+
+    return survey_match.group(2).decode()
+
```

Comment:
:+1: makes sense, though it should be the `upload_ztf_bytes_to_bucket` function that gets renamed, as that is the entry point. 

If we make this change, we must also update the file at `broker/setup_broker/deploy_cloud_fncs.sh`.
It would be simplest for me to make this change myself, if we do it.

Reply:
In this case I'm not sure that renaming `upload_ztf_bytes_to_bucket` is the direction, but maybe just have a `run` that immediately calls `upload_ztf_bytes_to_bucket`.

It's probably still correct to consider updating `broker/setup_broker/deploy_cloud_fncs.sh`

Reply:
Agreed. In general, I would expect that the `upload_ztf_bytes_to_bucket` function shouldn't care about the schema, and should just try to upload the data. `run` should then internally call `fix_schema` followed by `upload_ztf_bytes_to_bucket`. However, I don't want to push too hard on refactoring since there is already a lot going on in this PR.

Reply:
> Agreed. In general, I would expect that the `upload_ztf_bytes_to_bucket` function shouldn't care about the schema, and should just try to upload the data. `run` should then internally call `fix_schema` followed by `upload_ztf_bytes_to_bucket`.

That logic would be cleaner, but those two functions are tied together in a `with` that holds the file object open. I know this could be separated, I just don't want to start messing around with it right now. This component of the broker runs smoothly with very little lag, even if you flood it with alerts. If I were to work on this function, the first thing I would do is to add a `try/except` and see if I could get the packet loss to near 0 (currently, roughly 0.5% of alerts are dropped without getting into the bucket).

> However, I don't want to push too hard on refactoring since there is already a lot going on in this PR.

That's my vote; would rather work on other things.

Reply:
> In this case I'm not sure that renaming `upload_ztf_bytes_to_bucket` is the direction, but maybe just have a `run` that immediately calls `upload_ztf_bytes_to_bucket`.
> 
> It's probably still correct to consider updating `broker/setup_broker/deploy_cloud_fncs.sh`

This is simple, I will do it.